cvmfs/loader.cc

```diff
@@ -179,6 +179,77 @@ static void Usage(const string &exename) {
      exename.c_str());
 }
 
+void UmountOnCleanup(const string* mountpoint_) {
+  if (!mountpoint_) {
+    LogCvmfs(kLogCvmfs, kLogSyslogErr, "crash cleanup handler: no mountpoint");
+    return;
+  }
+
+  std::vector<std::string> all_mountpoints = platform_mountlist();
+  if (all_mountpoints.empty()) {
+    LogCvmfs(kLogCvmfs, kLogSyslogErr,
+             "crash cleanup handler: "
+             "failed to read mount point list");
+    return;
+  }
+
+  // Mitigate auto-mount - crash - umount - auto-mount loops
+  SafeSleepMs(2000);
+
+  // Check if *mountpoint_ is still mounted
+  // (we don't want to trigger a mount by immediately doing stat *mountpoint_)
+  bool still_mounted = false;
+  for (unsigned i = 0; i < all_mountpoints.size(); ++i) {
+    if (*mountpoint_ == all_mountpoints[i]) {
+      still_mounted = true;
+      break;
+    }
+  }
+  if (!still_mounted) {
+    LogCvmfs(kLogCvmfs, kLogSyslog, "crash cleanup handler: %s not mounted",
+             mountpoint_->c_str());
+    return;
+  }
+
+  // stat() might be served from caches.  Opendir ensures fuse module is called.
+  int expected_error;
+#ifdef __APPLE__
+  expected_error = ENXIO;
+#else
+  expected_error = ENOTCONN;
+#endif
+  DIR *dirp = opendir(mountpoint_->c_str());
+  if (dirp || (errno != expected_error)) {
+    if (dirp)
+      closedir(dirp);
+    LogCvmfs(kLogCvmfs, kLogSyslog,
+             "unmount un cleanup: "
+             "%s seems not to be stalled (%d)",
+             mountpoint_->c_str(), errno);
+    return;
+  }
+
+  // sudo umount -l *mountpoint_
+  if (!SwitchCredentials(0, getegid(), true)) {
```

Comment:
If you're doing the SwitchCredentials here then it doesn't also need to be done before calling this function.